src/views/projects/Cob/Revision.svelte

```diff
@@ -0,0 +1,270 @@
+<script lang="ts">
+  import type { BaseUrl, CommitHeader, Diff, Patch } from "@httpd-client";
+  import type { DiffResponse } from "@httpd-client/lib/project";
+  import type { Timeline } from "@app/views/projects/Patch.svelte";
+
+  import * as utils from "@app/lib/utils";
+  import { HttpdClient } from "@httpd-client";
+  import { sessionStore } from "@app/lib/session";
+  import { onMount } from "svelte";
+
+  import Authorship from "@app/components/Authorship.svelte";
+  import Avatar from "@app/components/Avatar.svelte";
+  import Clipboard from "@app/components/Clipboard.svelte";
+  import CommentComponent from "@app/components/Comment.svelte";
+  import DiffStatBadge from "@app/components/DiffStatBadge.svelte";
+  import ErrorMessage from "@app/components/ErrorMessage.svelte";
+  import Icon from "@app/components/Icon.svelte";
+  import ProjectLink from "@app/components/ProjectLink.svelte";
+  import ThreadComponent from "@app/components/Thread.svelte";
+
+  export let baseUrl: BaseUrl;
+  export let expanded: boolean = true;
+  export let patch: Patch;
+  export let projectHead: string;
+  export let projectId: string;
+  export let timelines: Timeline[];
+  export let revision: {
+    id: string;
+    oid: string;
+    base: string;
+    timestamp: number;
+    diff: Promise<DiffResponse>;
```

Comment:
Yeah I'm also not 100% sold on it, but I felt like it's the cleanest way, the other implementation I did was to do the diff resolving before when creating the timeline but it was more cumbersome to resolve the promises inside the timeline elements and catching and handling the error is even worse.

Reply:
Ok this commit avoids passing the promise to `Revision` since it resolves it on mounting `Revision`
f4e3ae9

Reply:
The downside is that we loose the stackTrace on why the revision diff failed.. But I'll think about an improvement.

Reply:
Maybe we can move the fetching itself back to the `Revision` component and just pass in the params instead of a promise?

```diff
diff --git a/src/views/projects/Cob/Revision.svelte b/src/views/projects/Cob/Revision.svelte
index 21dcda76..c8c84d3a 100644
--- a/src/views/projects/Cob/Revision.svelte
+++ b/src/views/projects/Cob/Revision.svelte
@@ -2,7 +2,9 @@
   import type { BaseUrl, DiffResponse } from "@httpd-client";
   import type { Timeline } from "@app/views/projects/Patch.svelte";
 
+  import { onMount } from "svelte";
   import * as utils from "@app/lib/utils";
+  import { HttpdClient } from "@httpd-client";
 
   import Authorship from "@app/components/Authorship.svelte";
   import Avatar from "@app/components/Avatar.svelte";
@@ -21,9 +23,13 @@
   export let authorId: string;
   export let projectHead: string;
   export let projectId: string;
-  export let diff: DiffResponse | undefined;
   export let timelines: Timeline[];
-  export let revision: { id: string; timestamp: number };
+  export let revisionId: string;
+  export let revisionTimestamp: number;
+  export let revisionBase: string;
+  export let revisionOid: string;
+
+  const api = new HttpdClient(baseUrl);
 
   function formatVerdict(revision: string, verdict?: string | null) {
     switch (verdict) {
@@ -35,6 +41,21 @@
         return `left a comment on revision ${utils.formatObjectId(revision)}`;
     }
   }
+
+  let response: DiffResponse | undefined = undefined;
+  let error: any | undefined = undefined;
+
+  onMount(async () => {
+    try {
+      response = await api.project.getDiff(
+        projectId,
+        revisionBase,
+        revisionOid,
+      );
+    } catch (err: any) {
+      error = err;
+    }
+  });
 </script>
 
 <style>
@@ -115,18 +136,18 @@
           on:click={() => (expanded = !expanded)} />
       </div>
       <span>
-        Revision {utils.formatObjectId(revision.id)}
+        Revision {utils.formatObjectId(revisionId)}
       </span>
-      <Clipboard text={revision.id} small />
+      <Clipboard text={revisionId} small />
     </div>
     <div class="txt-small" />
     <div class="revision-data">
-      {#if diff?.diff.stats}
-        {@const { insertions, deletions } = diff.diff.stats}
+      {#if response?.diff.stats}
+        {@const { insertions, deletions } = response.diff.stats}
         <DiffStatBadge {insertions} {deletions} />
       {/if}
       <span class="layout-desktop txt-small">
-        {utils.formatTimestamp(revision.timestamp)}
+        {utils.formatTimestamp(revisionTimestamp)}
       </span>
     </div>
   </div>
@@ -157,9 +178,9 @@
               </Authorship>
             </div>
           {/if}
-          {#if diff?.commits}
+          {#if response?.commits}
             <div class="action txt-tiny">
-              {#each diff.commits as commit}
+              {#each response.commits as commit}
                 <div class="commit-event">
                   <span>
                     <Avatar inline nodeId={authorId} />
@@ -180,9 +201,12 @@
                 </div>
               {/each}
             </div>
-          {:else}
+          {/if}
+          {#if error}
             <div class="txt-monospace">
-              <ErrorMessage message="Failed to load diff for this revision." />
+              <ErrorMessage
+                message="Failed to load diff for this revision."
+                stackTrace={error?.stack.toString()} />
             </div>
           {/if}
         {:else if element.type === "merge"}
diff --git a/src/views/projects/Patch.svelte b/src/views/projects/Patch.svelte
index 3bf17355..bfa88bcf 100644
--- a/src/views/projects/Patch.svelte
+++ b/src/views/projects/Patch.svelte
@@ -38,7 +38,7 @@
 </script>
 
 <script lang="ts">
-  import type { BaseUrl, Patch, DiffResponse } from "@httpd-client";
+  import type { BaseUrl, Patch } from "@httpd-client";
   import type { Variant } from "@app/components/Badge.svelte";
 
   import * as utils from "@app/lib/utils";
@@ -132,12 +132,14 @@
 
   const currentRevision =
     patch.revisions.find(r => r.id === revision) || patch.revisions[0];
+
   $: timelineTuple = patch.revisions.map<
     [
       {
         id: string;
         timestamp: number;
-        diff: Promise<DiffResponse | undefined>;
+        revisionBase: string;
+        revisionOid: string;
       },
       Timeline[],
     ]
@@ -145,9 +147,8 @@
     {
       id: rev.id,
       timestamp: rev.timestamp,
-      diff: api.project
-        .getDiff(projectId, rev.base, rev.oid)
-        .catch(() => undefined),
+      revisionBase: rev.base,
+      revisionOid: rev.oid,
     },
     [
       ...rev.reviews.map<TimelineReview>(([author, review]) => ({
@@ -325,19 +326,19 @@
     </div>
     {#if currentTab === "activity"}
       {#each timelineTuple as [revision, timelines], index}
-        {#await revision.diff then diff}
-          <RevisionComponent
-            {baseUrl}
-            {projectId}
-            {timelines}
-            {projectHead}
-            {revision}
-            {diff}
-            on:reply={createReply}
-            patchId={patch.id}
-            authorId={patch.author.id}
-            expanded={index === patch.revisions.length - 1} />
-        {/await}
+        <RevisionComponent
+          {baseUrl}
+          {projectId}
+          {timelines}
+          {projectHead}
+          revisionId={revision.id}
+          revisionTimestamp={revision.timestamp}
+          revisionBase={revision.revisionBase}
+          revisionOid={revision.revisionOid}
+          on:reply={createReply}
+          patchId={patch.id}
+          authorId={patch.author.id}
+          expanded={index === patch.revisions.length - 1} />
       {:else}
         <Placeholder emoji="🍂">
           <div slot="title">No activity</div>

```

Reply:
nice solution thanks!, addressed in fd6a6fd66dc312fe8c7e907aa8b5d4054993fdc4

Reply:
Will add you as a co-author